File: polb/fln.f
*
* Copyright (c) 2020 NVI, Inc.
*
* This file is part of VLBI Field System
* (see http://github.com/nvi-inc/fs).
*
* This program is free software: you can redistribute it and/or modify
* it under the terms of the GNU General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* This program is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU General Public License for more details.
*
* You should have received a copy of the GNU General Public License
* along with this program. If not, see <http://www.gnu.org/licenses/>.
*
      double precision function fln(iwhich,x,y,p,ipar,phi)
C
C     FITTING FUNCTION FOR X COORDINATE INCLUDING DERIVATIVES
C     FUNCTION FOR IWHICH = 0
C     DERIVATIVE WITH RESPECT TO THE (IWHICH)th PARAMETER OTHERWISE
C     COULD HANDLE UP TO 20 PARAMETERS
C
      dimension ipar(20)
      double precision f,cosx,cosy,cosl,sinx,siny,sinl,x,y,p(20),phi
      double precision sin2x,cos2x
C
      cosx=dcos(x)
      cosy=dcos(y)
      cosl=dcos(phi)
      sinx=dsin(x)
      siny=dsin(y)
      sinl=dsin(phi)
      sin2x=dsin(2.0d0*x)
      cos2x=dcos(2.0d0*x)
      f=0.0d0
C
      if (iwhich.lt.0.or.iwhich.gt.20) goto 1000
      goto (1,  10,  20,  30,  40,  50,
     +           60,1000,1000,1000,1000,
     +         1000, 120, 130, 140,1000,
     +         1000, 170, 180,1000,1000) iwhich+1
C
1     continue
      if (ipar( 1).ne.0) f=f+p(1)
      if (ipar( 2).ne.0) f=f-p(2)*cosl*sinx/cosy
      if (ipar( 3).ne.0) f=f+p(3)*siny/cosy
      if (ipar( 4).ne.0) f=f-p(4)/cosy
      if (ipar( 5).ne.0) f=f+p(5)*sinx*siny/cosy
      if (ipar( 6).ne.0) f=f-p(6)*siny*cosx/cosy
      if (ipar(12).ne.0) f=f+p(12)*x
      if (ipar(13).ne.0) f=f+p(13)*cosx
      if (ipar(14).ne.0) f=f+p(14)*sinx
      if (ipar(17).ne.0) f=f+p(17)*cos2x
      if (ipar(18).ne.0) f=f+p(18)*sin2x
      goto 1000
C
10    continue
      if (ipar(1).ne.0) f=1.0d0
      goto 1000
C
20    continue
      if (ipar(2).ne.0) f=-cosl*sinx/cosy
      goto 1000
C
30    continue
      if (ipar(3).ne.0) f= siny/cosy
      goto 1000
C
40    continue
      if (ipar(4).ne.0) f=-1.0d0/cosy
      goto 1000
C
50    continue
      if (ipar(5).ne.0) f= (sinx*siny)/cosy
      goto 1000
C
60    continue
      if (ipar(6).ne.0) f=-(cosx*siny)/cosy
      goto 1000
120   continue
      if (ipar(12).ne.0) f=x
      goto 1000
130   continue
      if (ipar(13).ne.0) f=cosx
      goto 1000
140   continue
      if (ipar(14).ne.0) f=sinx
      goto 1000
170   continue
      if (ipar(17).ne.0) f=cos2x
      goto 1000
180   continue
      if (ipar(18).ne.0) f=sin2x
      goto 1000
C
1000  continue
      fln=f

      return
      end

File: poclb/fln.c
/*
 * Copyright (c) 2020 NVI, Inc.
 *
 * This file is part of VLBI Field System
 * (see http://github.com/nvi-inc/fs).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
/* fln

   Fitting function for X coordinate including derivatives.
   Derivative with respect to the (iwitch)th parameter.
   Copied from FORTRAN.

*/

#include <math.h>
#include "../include/pmodel.h"

double fln(iwhich,x,y,pmodel)
int iwhich;
double x,y;
struct pmdl *pmodel;

{
  double cosx,cosy,cosl,sinx,siny,sinl,f,cos2x,sin2x;

  cosx = cos(x);
  cosy = cos(y);
  cosl = cos(pmodel->phi);
  sinx = sin(x);
  siny = sin(y);
  cos2x = cos(2.*x);
  sin2x = sin(2.*x);
  sinl = sin(pmodel->phi);
  f = 0.0;

  switch (iwhich-1) {
    case -1:
      if (pmodel->ipar[ 0] != 0) f=f+pmodel->pcof[0];
      if (pmodel->ipar[ 1] != 0) f=f-pmodel->pcof[1]*cosl*sinx/cosy;
      if (pmodel->ipar[ 2] != 0) f=f+pmodel->pcof[2]*siny/cosy;
      if (pmodel->ipar[ 3] != 0) f=f-pmodel->pcof[3]/cosy;
      if (pmodel->ipar[ 4] != 0) f=f+pmodel->pcof[4]*sinx*siny/cosy;
      if (pmodel->ipar[ 5] != 0) f=f-pmodel->pcof[5]*siny*cosx/cosy;
      if (pmodel->ipar[11] != 0) f=f+pmodel->pcof[11]*x;
      if (pmodel->ipar[12] != 0) f=f+pmodel->pcof[12]*cosx;
      if (pmodel->ipar[13] != 0) f=f+pmodel->pcof[13]*sinx;
      if (pmodel->ipar[16] != 0) f=f+pmodel->pcof[16]*cos2x;
      if (pmodel->ipar[17] != 0) f=f+pmodel->pcof[17]*sin2x;
      break;
    case 0:
      if (pmodel->ipar[0] != 0) f=1.0;
      break;
    case 1:
      if (pmodel->ipar[1] != 0) f=-cosl*sinx/cosy;
      break;
    case 2:
      if (pmodel->ipar[2] != 0) f= siny/cosy;
      break;
    case 3:
      if (pmodel->ipar[3] != 0) f=-1.0/cosy;
      break;
    case 4:   
      if (pmodel->ipar[4] != 0) f= (sinx*siny)/cosy;
      break;
    case 5:   
      if (pmodel->ipar[5] != 0) f=-(cosx*siny)/cosy;
      break;
    case 11:   
      if (pmodel->ipar[11] != 0) f=x;
      break;
    case 12:  
      if (pmodel->ipar[12] != 0) f=cosx;
      break;
    case 13:
      if (pmodel->ipar[13] != 0) f=sinx;
      break;
    case 16:
      if (pmodel->ipar[16] != 0) f=cos2x;
      break;
    case 17:
      if (pmodel->ipar[17] != 0) f=sin2x;
      break;
  default:
    f=0.0;
    break;
 } 
    return f;
}

